src/connectdlg.cpp

```diff
@@ -411,8 +411,11 @@ void CConnectDlg::SetServerList ( const CHostAddress& InetAddr, const CVector<CS
 
         if ( vecServerInfo[iIdx].eCountry != QLocale::AnyCountry )
         {
+#if QT_VERSION >= QT_VERSION_CHECK( 6, 0, 0 )
+            QString strCountryToString = QLocale::countryToString ( CLocale::LocaleQt5toQt6 ( vecServerInfo[iIdx].eCountry ) );
```

Comment:
Hm.  Reading this now raises concerns.

We need backwards compatibility -- but I don't think this is the right way to achieve it.  The content of `vecServerInfo[iIdx].eCountry` should be Qt6-compatible at this point in the code, so however that vec is populated should be doing the translate.

I think that would mean fewer changes to the code.

Reply:
So, the origin of `vecServerInfo` is:
CProtocol::CLServerListReceived -> CClient::CLServerListReceived -> CConnectDlg::OnCL(Red)ServerListReceived -> CConnectDlg::SetServerList

Not sure if it'll really be fewer changes, but I agree that it would be more correct to do it on the lowest possible level.

@pljones So I guess the best place would be protocol.cpp, so that the contract would be that the caller always puts in / gets out a value which is valid for the currently used Qt version, right?

Reply:
I have pushed that now.
Old: https://github.com/jamulussoftware/jamulus/pull/2299/commits/4292008f78f553d37a71d56be42d9d64180d2266
New: https://github.com/jamulussoftware/jamulus/pull/2299/commits/0c6055cde8443215eb4f201c2999472c78de118b

So, the new approach (protocol layer handling) requires more changes. However:
- It handles more cases (the previous approach seemed like it didn't handle the Server UI, for example)
- It includes the helper script which updates the code translation table in order to make that reproducible and maintainable.
- I believe the new approach as suggested by you (@pljones) is better at using the right layer of abstraction instead of having to think about this in all the places where this is used.

There's one caveat though: The country flag handling currently hardcodes a code to flag name mapping. According to the comments, this is due to Qt4 compatibility. The code already includes code which should work for Qt5+. So, for Qt6 support in the flag handling we've got two options:
- The protocol layer gives us Qt-native (i.e. Qt6 in Qt6) country codes now. As the hardcoded `switch` uses Qt5 country codes, we would have to convert *back*. I've currently implemented this approach because it limits the changes to the Qt6-specific `#ifdef`.
- We could drop the hardcoding (I don't think we are Qt4-compatible anymore anyway?) in favor of Volker's Qt-native code which should (and seems to) work with both Qt5 and Qt6. I think this is the better solution. I'm just unsure if we should do that within this PR or not.

Reply:
We no longer support build on a Qt 4 environment.  However, for clarity, does the comment mean that or is it talking about something else, like a Qt4 client talking (via a server) to a Qt5 client?

Reply:
I'm pretty sure it's about the implementation. To be extra safe, I'm going to build a test case which compares the output of the commented out logic with the hardcoded values.

Reply:
**Test environment**
Special Jamulus builds for Qt5 and Qt6 based on current Qt6-compat branch from this PR with a patch to output code+flag for all supported country codes. Details:
<details>
<summary>Setup instructions/Patch</summary>

```
$ git checkout 9207e341ccc34be93ebde130d44282b3cbdad9fd
$ patch -p1
diff --git a/src/main.cpp b/src/main.cpp
index 3ae7aab2..f2b8a2e7 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -50,6 +50,27 @@ extern void qt_set_sequence_auto_mnemonic ( bool bEnable );
 
 int main ( int argc, char** argv )
 {
+    bool bUseTable = argc > 1 ? std::string ( argv[1] ) != "qt" : true;
+    qInfo() << "qt version:" << QT_VERSION_STR;
+    qInfo() << "mode:" << ( bUseTable ? "table" : "qt" );
+    for ( int iCurCntry = static_cast<int> ( QLocale::AnyCountry ); iCurCntry < static_cast<int> ( QLocale::LastCountry ); iCurCntry++ )
+    {
+        if ( static_cast<QLocale::Country> ( iCurCntry ) == QLocale::AnyCountry )
+        {
+            // continue;
+        }
+
+        if ( !CLocale::IsCountryCodeSupported ( iCurCntry ) )
+        {
+            continue;
+        }
+
+        qInfo() << "code" << iCurCntry << "-" /*<< QLocale::countryToString ( static_cast<QLocale::Country> ( iCurCntry ) ) << " - " */
+                << CLocale::GetCountryFlagIconsResourceReference (
+                       CLocale::QtCountryToWireFormatCountry ( static_cast<QLocale::Country> ( iCurCntry ) ),
+                       bUseTable );
+    }
+    exit ( 0 );
 
 #if defined( Q_OS_MACX )
     // Mnemonic keys are default disabled in Qt for MacOS. The following function enables them.
diff --git a/src/util.cpp b/src/util.cpp
index 364dd86e..f442272c 100644
--- a/src/util.cpp
+++ b/src/util.cpp
@@ -1246,7 +1246,7 @@ bool CLocale::IsCountryCodeSupported ( uint iCountryCode )
 #endif
 }
 
-QString CLocale::GetCountryFlagIconsResourceReference ( const QLocale::Country eCountry /* Always a Qt5 (!) code */ )
+QString CLocale::GetCountryFlagIconsResourceReference ( const QLocale::Country eCountry /* Always a Qt5 (!) code */, bool bUseTable )
 {
     QString strReturn = "";
 
@@ -1255,6 +1255,9 @@ QString CLocale::GetCountryFlagIconsResourceReference ( const QLocale::Country e
     {
         return ":/png/flags/res/flags/flagnone.png";
     }
+
+    if ( bUseTable )
+    {
         // NOTE: The following code was introduced to support old QT versions. The problem
         //       is that the number of countries displayed is less than the one displayed
         //       with the new code below (which is disabled). Therefore, as soon as the
@@ -1827,15 +1830,15 @@ QString CLocale::GetCountryFlagIconsResourceReference ( const QLocale::Country e
             return "";
         }
         return strReturn;
-
+    }
+    else
+    {
         // AT LEAST QT 4.8 IS REQUIRED:
-    /*
+
         // There is no direct query of the country code in Qt, therefore we use a
         // workaround: Get the matching locales properties and split the name of
         // that since the second part is the country code
-    QList<QLocale> vCurLocaleList = QLocale::matchingLocales ( QLocale::AnyLanguage,
-            QLocale::AnyScript,
-            eCountry );
+        QList<QLocale> vCurLocaleList = QLocale::matchingLocales ( QLocale::AnyLanguage, QLocale::AnyScript, eCountry );
 
         // check if the matching locales query was successful
         if ( vCurLocaleList.size() > 0 )
@@ -1860,7 +1863,7 @@ QString CLocale::GetCountryFlagIconsResourceReference ( const QLocale::Country e
             }
         }
         return strReturn;
-    */
+    }
 }
 
 QMap<QString, QString> CLocale::GetAvailableTranslations()
diff --git a/src/util.h b/src/util.h
index 56bb6ff6..413b4552 100644
--- a/src/util.h
+++ b/src/util.h
@@ -809,7 +809,7 @@ protected:
 class CLocale
 {
 public:
-    static QString                 GetCountryFlagIconsResourceReference ( const QLocale::Country eCountry /* Always a Qt5 (!) code */ );
+    static QString GetCountryFlagIconsResourceReference ( const QLocale::Country eCountry /* Always a Qt5 (!) code */, bool bUseTable = true );
     static QMap<QString, QString>  GetAvailableTranslations();
     static QPair<QString, QString> FindSysLangTransFileName ( const QMap<QString, QString>& TranslMap );
     static void                    LoadTranslation ( const QString strLanguage, QCoreApplication* pApp );
^D

$ make distclean; qmake && make -j && mv Jamulus Jamulus-qt5
$ make distclean; qmake6 && make -j && mv Jamulus Jamulus-qt6
$ ./Jamulus-qt5 table &> /tmp/qt5-table
$ ./Jamulus-qt5 qt &> /tmp/qt5-qt
$ ./Jamulus-qt6 table &> /tmp/qt6-table
$ ./Jamulus-qt6 qt &> /tmp/qt6-qt
```
</details>

**Qt5: Table (existing) vs. `QLocale::matchingLocales`**
Result: `QLocale::matchingLocales` supports 53 more flags and doesn't lack a single one.
```
diff -u0 /tmp/qt5-{table,qt} | grep -v '@@' | grep -c '^-code'
53
```
<details>
<summary>diff -u0 /tmp/qt5-{table,qt} | grep -v '@@'</summary>

```diff
--- /tmp/qt5-table	2022-02-08 01:02:36.072798047 +0100
+++ /tmp/qt5-qt	2022-02-08 01:02:36.086131545 +0100
-mode: table
+mode: qt
-code 4 - ""
+code 4 - ":/png/flags/res/flags/as.png"
-code 7 - ""
+code 7 - ":/png/flags/res/flags/ai.png"
-code 9 - ""
+code 9 - ":/png/flags/res/flags/ag.png"
-code 16 - ""
+code 16 - ":/png/flags/res/flags/bs.png"
-code 19 - ""
+code 19 - ":/png/flags/res/flags/bb.png"
-code 22 - ""
+code 22 - ":/png/flags/res/flags/bz.png"
-code 24 - ""
+code 24 - ":/png/flags/res/flags/bm.png"
-code 31 - ""
+code 31 - ":/png/flags/res/flags/io.png"
-code 40 - ""
+code 40 - ":/png/flags/res/flags/ky.png"
-code 45 - ""
-code 46 - ""
+code 45 - ":/png/flags/res/flags/cx.png"
+code 46 - ":/png/flags/res/flags/cc.png"
-code 51 - ""
+code 51 - ":/png/flags/res/flags/ck.png"
-code 60 - ""
+code 60 - ":/png/flags/res/flags/dm.png"
-code 70 - ""
+code 70 - ":/png/flags/res/flags/fk.png"
-code 72 - ""
+code 72 - ":/png/flags/res/flags/fj.png"
-code 80 - ""
+code 80 - ":/png/flags/res/flags/gm.png"
-code 84 - ""
+code 84 - ":/png/flags/res/flags/gi.png"
-code 87 - ""
+code 87 - ":/png/flags/res/flags/gd.png"
-code 89 - ""
+code 89 - ":/png/flags/res/flags/gu.png"
-code 94 - ""
+code 94 - ":/png/flags/res/flags/ht.png"
-code 107 - ""
+code 107 - ":/png/flags/res/flags/jm.png"
-code 112 - ""
+code 112 - ":/png/flags/res/flags/ki.png"
-code 121 - ""
+code 121 - ":/png/flags/res/flags/lr.png"
-code 129 - ""
+code 129 - ":/png/flags/res/flags/mw.png"
-code 131 - ""
+code 131 - ":/png/flags/res/flags/mv.png"
-code 134 - ""
+code 134 - ":/png/flags/res/flags/mh.png"
-code 140 - ""
+code 140 - ":/png/flags/res/flags/fm.png"
-code 144 - ""
+code 144 - ":/png/flags/res/flags/ms.png"
-code 149 - ""
+code 149 - ":/png/flags/res/flags/nr.png"
-code 158 - ""
-code 159 - ""
-code 160 - ""
+code 158 - ":/png/flags/res/flags/nu.png"
+code 159 - ":/png/flags/res/flags/nf.png"
+code 160 - ":/png/flags/res/flags/mp.png"
-code 164 - ""
+code 164 - ":/png/flags/res/flags/pw.png"
-code 171 - ""
+code 171 - ":/png/flags/res/flags/pn.png"
-code 180 - ""
-code 181 - ""
-code 182 - ""
-code 183 - ""
+code 180 - ":/png/flags/res/flags/kn.png"
+code 181 - ":/png/flags/res/flags/lc.png"
+code 182 - ":/png/flags/res/flags/vc.png"
+code 183 - ":/png/flags/res/flags/ws.png"
-code 193 - ""
+code 193 - ":/png/flags/res/flags/sb.png"
-code 199 - ""
-code 200 - ""
+code 199 - ":/png/flags/res/flags/sh.png"
+code 200 - ":/png/flags/res/flags/pm.png"
-code 203 - ""
+code 203 - ":/png/flags/res/flags/sj.png"
-code 213 - ""
+code 213 - ":/png/flags/res/flags/tk.png"
-code 215 - ""
+code 215 - ":/png/flags/res/flags/tt.png"
-code 218 - ""
-code 219 - ""
-code 220 - ""
+code 218 - ":/png/flags/res/flags/tm.png"
+code 219 - ":/png/flags/res/flags/tc.png"
+code 220 - ":/png/flags/res/flags/tv.png"
-code 226 - ""
+code 226 - ":/png/flags/res/flags/um.png"
-code 229 - ""
-code 230 - ""
+code 229 - ":/png/flags/res/flags/vu.png"
+code 230 - ":/png/flags/res/flags/va.png"
-code 233 - ""
-code 234 - ""
-code 235 - ""
+code 233 - ":/png/flags/res/flags/vg.png"
+code 234 - ":/png/flags/res/flags/vi.png"
+code 235 - ":/png/flags/res/flags/wf.png"
```
</details>

**Qt6: Table (existing + Qt6-back-translation in 9207e341ccc34be93ebde130d44282b3cbdad9fd) vs. `QLocale::matchingLocales`**
Result: `QLocale::matchingLocales` supports 53 more flags and doesn't lack a single one.
```
diff -u0 /tmp/qt6-{table,qt} | grep -v '@@' | grep -c '^-code'
53
```
<details>
<summary>diff -u0 /tmp/qt6-{table,qt} | grep -v '@@'</summary>

```diff
--- /tmp/qt6-table	2022-02-08 01:02:36.096131669 +0100
+++ /tmp/qt6-qt	2022-02-08 01:02:36.106131793 +0100
-mode: table
+mode: qt
-code 6 - ""
+code 6 - ":/png/flags/res/flags/as.png"
-code 9 - ""
+code 9 - ":/png/flags/res/flags/ai.png"
-code 11 - ""
+code 11 - ":/png/flags/res/flags/ag.png"
-code 14 - ""
+code 14 - ":/png/flags/res/flags/um.png"
-code 20 - ""
+code 20 - ":/png/flags/res/flags/bs.png"
-code 23 - ""
+code 23 - ":/png/flags/res/flags/bb.png"
-code 26 - ""
+code 26 - ":/png/flags/res/flags/bz.png"
-code 28 - ""
+code 28 - ":/png/flags/res/flags/bm.png"
-code 36 - ""
-code 37 - ""
+code 36 - ":/png/flags/res/flags/io.png"
+code 37 - ":/png/flags/res/flags/vg.png"
-code 51 - ""
-code 52 - ""
+code 51 - ":/png/flags/res/flags/ky.png"
+code 52 - ":/png/flags/res/flags/tc.png"
-code 58 - ""
+code 58 - ":/png/flags/res/flags/cx.png"
-code 65 - ""
-code 66 - ""
+code 65 - ":/png/flags/res/flags/ck.png"
+code 66 - ":/png/flags/res/flags/vi.png"
-code 75 - ""
+code 75 - ":/png/flags/res/flags/dm.png"
-code 78 - ""
+code 78 - ":/png/flags/res/flags/zw.png"
-code 89 - ""
+code 89 - ":/png/flags/res/flags/fk.png"
-code 91 - ""
+code 91 - ":/png/flags/res/flags/fj.png"
-code 98 - ""
+code 98 - ":/png/flags/res/flags/gm.png"
-code 103 - ""
+code 103 - ":/png/flags/res/flags/gi.png"
-code 106 - ""
+code 106 - ":/png/flags/res/flags/gd.png"
-code 108 - ""
+code 108 - ":/png/flags/res/flags/gu.png"
-code 114 - ""
+code 114 - ":/png/flags/res/flags/ht.png"
-code 118 - ""
+code 118 - ":/png/flags/res/flags/cc.png"
-code 121 - ""
+code 121 - ":/png/flags/res/flags/vu.png"
-code 132 - ""
+code 132 - ":/png/flags/res/flags/jm.png"
-code 138 - ""
+code 138 - ":/png/flags/res/flags/ki.png"
-code 147 - ""
+code 147 - ":/png/flags/res/flags/lr.png"
-code 155 - ""
+code 155 - ":/png/flags/res/flags/mw.png"
-code 157 - ""
-code 158 - ""
+code 157 - ":/png/flags/res/flags/tv.png"
+code 158 - ":/png/flags/res/flags/mv.png"
-code 161 - ""
+code 161 - ":/png/flags/res/flags/mh.png"
-code 166 - ""
+code 166 - ":/png/flags/res/flags/fm.png"
-code 171 - ""
+code 171 - ":/png/flags/res/flags/ms.png"
-code 178 - ""
+code 178 - ":/png/flags/res/flags/nr.png"
-code 186 - ""
-code 187 - ""
-code 188 - ""
+code 186 - ":/png/flags/res/flags/nu.png"
+code 187 - ":/png/flags/res/flags/nf.png"
+code 188 - ":/png/flags/res/flags/mp.png"
-code 194 - ""
+code 194 - ":/png/flags/res/flags/pw.png"
-code 200 - ""
+code 200 - ":/png/flags/res/flags/pm.png"
-code 205 - ""
+code 205 - ":/png/flags/res/flags/pn.png"
-code 211 - ""
+code 211 - ":/png/flags/res/flags/wf.png"
-code 217 - ""
+code 217 - ":/png/flags/res/flags/sh.png"
-code 220 - ""
-code 221 - ""
-code 222 - ""
-code 223 - ""
+code 220 - ":/png/flags/res/flags/kn.png"
+code 221 - ":/png/flags/res/flags/lc.png"
+code 222 - ":/png/flags/res/flags/vc.png"
+code 223 - ":/png/flags/res/flags/ws.png"
-code 235 - ""
+code 235 - ":/png/flags/res/flags/sb.png"
-code 237 - ""
+code 237 - ":/png/flags/res/flags/va.png"
-code 246 - ""
+code 246 - ":/png/flags/res/flags/sj.png"
-code 249 - ""
+code 249 - ":/png/flags/res/flags/tk.png"
-code 257 - ""
+code 257 - ":/png/flags/res/flags/tt.png"
```
</details>

**Is a single flag image used for multiple codes?**
No:
```
grep -oP '":.*"' /tmp/qt* | sort | uniq -c | grep -v '1 '
```


**Is any flag from the reference (Qt5, table) missing in any other scenario (Qt5/QLocale, Qt6/Table, Qt6/Qlocale)?**
```
$ for flag in $(grep -oP '":.*"' /tmp/qt5-table | sort); do for scenario in /tmp/qt5-qt /tmp/qt6-table /tmp/qt6-qt; do grep -qF "$flag" "$scenario" || echo "$scenario is missing $flag"; done; done
/tmp/qt6-table is missing ":/png/flags/res/flags/zw.png"
```
- [ ] This will have to be investigated as it might point towards a problem in the code translation table. Looks like the reason might be that Qt5 uses `240` for two different entries. Should probably adjust the logic to last-one-wins to work around that?
```
$ grep 240 /usr/include/qt/QtCore/qlocale.h
        Duala = 240,
        Zimbabwe = 240,
$ grep Zimbabwe /usr/include/qt6/QtCore/qlocale.h
        Zimbabwe = 261,
        LastTerritory = Zimbabwe,
$ grep Duala /usr/include/qt6/QtCore/qlocale.h
        Duala = 71,
```
**Summary**
It looks like we could safely switch to the `QLocale::matchingLocales` approach as it delivers all of what the table delivers + 53 additional flags.
- [ ] We should check that those new flag assignments are correct.

Reply:
The issue for Zimbabwe missing was a off-by-one error in the translation table generation script and its output. After fixing this, the last check from the previous comment does not yield anything anymore (as expected) in 3ac8c17d2df0bec436f3f6c34e29c82771e7a5f6.

However, since switching to the `QLocale`-based lookups is a change in functionality (it adds 53 countries/flags!), I'm splitting this out into its own PR.

Reply:
The flag-related change is now split into #2409. Will set this PR to draft (also for the unresolved settings stuff) and will rebase/continue once #2409 is done.

Reply:
With #2409 merged and this PR rebased, I'm marking this Flag-related thread as resolved. It also cleared out several other review comments as the suggestions were around code which has been eliminated by now. :)